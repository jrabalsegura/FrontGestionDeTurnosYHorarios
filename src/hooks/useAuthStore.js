import { useDispatch, useSelector } from "react-redux"
import gestionApi from "../api/gestionApi"
import { onChecking, onLogin, onLogout, clearErrorMessage, onUpdateUser } from "../store"


export const useAuthStore = () => {

    const { status, user, errorMessage } = useSelector(state => state.auth);
    const dispatch = useDispatch();

    const startLogin = async ({ username, password }) => {
        dispatch(onChecking())
        console.log(username, password)
        //dispatch(login(username, password))

        try {
            const { data } = await gestionApi.post('/auth/', { username, password })
            console.log(data)
            localStorage.setItem('token', data.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(onLogin({name: data.name, uid: data.uid}))

        } catch (error) {
            dispatch(onLogout('Credenciales incorrectas'));
            setTimeout(() => {
                dispatch(clearErrorMessage())
            }, 10);
        }
    }

    const checkAuthToken = async () => {
        const token = localStorage.getItem('token');
        if(!token) return dispatch(onLogout());

        try {
            const { data } = await gestionApi.get('/auth/renew');
            if (data.ok) {
                localStorage.setItem('token', data.token);
                localStorage.setItem('token-init-date', new Date().getTime());
                dispatch(onLogin({name: data.name, uid: data.uid}))
            } else {
                dispatch(onLogout());
            }
        } catch (error) {
            localStorage.clear();
            dispatch(onLogout());
        }
    }

    const startLogout = () => {
        localStorage.clear();
        dispatch(onLogout());
    }

    const onUpdate = (user) => {
        dispatch(onUpdateUser(user));
    }

    return {
        status,
        user,
        errorMessage,
        startLogin,
        checkAuthToken,
        startLogout,
        onUpdate
    }
}

