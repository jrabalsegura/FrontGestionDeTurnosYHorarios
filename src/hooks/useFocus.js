import { useRef, useState, useCallback, useEffect } from 'react';

export const useFocus = () => {
    const ref = useRef(null)
    const [isFocused, setIsFocused] = useState(false)
    const toggle = useCallback(() => {
      setIsFocused(!isFocused)
    }, [isFocused])
  
    useEffect(() => {
      const element = ref.current
  
      element?.addEventListener('focus', toggle)
      element?.addEventListener('blur', toggle)
  
      return () => {
        element?.removeEventListener('focus', toggle)
        element?.removeEventListener('blur', toggle)
      }
    }) 
  
    return [ref, isFocused]
  }