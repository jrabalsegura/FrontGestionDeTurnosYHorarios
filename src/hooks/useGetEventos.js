import { useState, useEffect } from "react";
import gestionApi from "../api/gestionApi";
import Swal from "sweetalert2";

export const useGetEventos = () => {
    const [events, setEvents] = useState([]);
    const [isLoading, setIsLoading] = useState(true);
    const [hasError, setHasError] = useState(false);

    useEffect(() => {
        const fetchEvents = async () => {
            try {
                const response = await gestionApi.get('/eventosTrabajo');
                const events = response.data.eventos;
                setEvents(events);
            } catch (error) {
                setHasError(true);
                setIsLoading(false);
                console.log(error);
                Swal.fire('Error al intentar obtener los eventos', error.message, 'error')
            }
        };
        fetchEvents();
    }, []);

    return {
        events,
        isLoading,
        hasError,
        setEvents
    }
}
