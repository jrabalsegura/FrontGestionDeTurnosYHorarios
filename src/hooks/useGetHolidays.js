import { useState, useEffect } from "react";
import gestionApi from "../api/gestionApi";
import Swal from "sweetalert2";

export const useGetHolidays = ({ id }) => {
    const [holidays, setHolidays] = useState([]);
    const [isLoading, setIsLoading] = useState(false);
    const [hasError, setHasError] = useState(false);

    useEffect(() => {
        const fetchHolidays = async () => {
            try {
                setIsLoading(true);
                const response = await gestionApi.get(`/holidays/${id}`);
                console.log(response)

                const adaptedHolidays = response.data.holidays.map(holiday => ({
                    title: 'Holidays',
                    start: new Date(holiday.startDate),
                    end: new Date(holiday.endDate),
                }));
                setHolidays(adaptedHolidays);
            } catch (error) {
                setHasError(true);
                console.log(error);
                Swal.fire('Error al intentar obtener las vacaciones', error.message, 'error')
            } finally {
                setIsLoading(false);
            }
        };
        fetchHolidays();
    }, []);

    return {
        holidays,
        isLoading,
        hasError,
        setHolidays
    }
}