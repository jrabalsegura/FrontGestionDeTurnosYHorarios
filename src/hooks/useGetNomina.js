import { useState, useEffect } from "react";
import gestionApi from "../api/gestionApi";
import Swal from "sweetalert2";

export const useGetNomina = ( {params} ) => {
    const [nomina, setNomina] = useState([null]);
    const [isLoading, setIsLoading] = useState(false);
    const [hasError, setHasError] = useState(false);

    useEffect(() => {
        const fetchNomina = async () => {
            try {
                setIsLoading(true);
                const response = await gestionApi.get(`/nominas`, {params});
                setNomina(response.data.nomina[0]);
            } catch (error) {
                setHasError(true);
                setNomina(null);
            } finally {
                setIsLoading(false);
            }
        };
        fetchNomina();
    }, [params.employeeId, params.month, params.year]);

    return {
        nomina,
        isLoading,
        hasError,
        setNomina
    }
}

