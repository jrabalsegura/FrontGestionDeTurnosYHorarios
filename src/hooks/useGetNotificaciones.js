import { useState, useEffect } from "react";
import gestionApi from "../api/gestionApi";
import Swal from "sweetalert2";
import { useSelector, useDispatch } from "react-redux";
import { onfetchedNotifications, removeNotificacion } from "../store/admin/notificacionesSlice";

export const useGetNotificaciones = () => {
    const { notificaciones } = useSelector((state) => state.notificaciones);
    const dispatch = useDispatch();
    const [isLoading, setIsLoading] = useState(false);
    const [hasError, setHasError] = useState(false);

    useEffect(() => {
        const fetchNotificaciones = async () => {
            try {
                setIsLoading(true);
                const response = await gestionApi.get('/notificaciones');
                dispatch(onfetchedNotifications(response.data.notifications));
            } catch (error) {
                setHasError(true);
                console.log(error);
                Swal.fire('Error al intentar obtener las notificaciones', error.message, 'error')
            } finally {
                setIsLoading(false);
            }
        };
        fetchNotificaciones();
    }, []);

    const filterNotificacion = (id) => {
        dispatch(removeNotificacion(id));
    }

    return {
        notificaciones,
        isLoading,
        hasError,
        filterNotificacion
    }
}

