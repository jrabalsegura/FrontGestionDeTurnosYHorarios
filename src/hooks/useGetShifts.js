import { useState, useEffect } from "react";
import gestionApi from "../api/gestionApi";
import Swal from "sweetalert2";

export const useGetShifts = ({ id }) => {
    const [shifts, setShifts] = useState([]);
    const [isLoading, setIsLoading] = useState(false);
    const [hasError, setHasError] = useState(false);

    useEffect(() => {
        const fetchShifts = async () => {
            try {
                setIsLoading(true);
                const response = await gestionApi.get(`/shifts/${id}`);

                const adaptedShifts = response.data.shifts.map(shift => ({
                    title: shift.type,
                    start: new Date(shift.start),
                    end: new Date(shift.end),
                }));
                setShifts(adaptedShifts);
            } catch (error) {
                setHasError(true);
                console.log(error);
                Swal.fire('Error al intentar obtener los turnos', error.message, 'error')
            } finally {
                setIsLoading(false);
            }
        };
        fetchShifts();
    }, []);

    return {
        shifts,
        isLoading,
        hasError,
        setShifts
    }
}
