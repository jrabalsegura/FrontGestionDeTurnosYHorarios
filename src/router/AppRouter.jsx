import { Navigate, Routes, Route } from 'react-router-dom';
import { AuthRoutes } from '../auth/routes/AuthRoutes';
import { AdminRoutes } from '../admin/routes/AdminRoutes';
import { EmployeeRoutes } from '../employee/routes/EmployeeRoutes';

export const AppRouter = () => {

    const authStatus = 'not-authenticated'

    return (
        <Routes>
            {
                (authStatus === 'not-authenticated') ? 
                    <Route path='/auth/*' element={<AuthRoutes />} /> : 

                    //If user.name === 'admin' go to admin routes
                    //else go to employee routes
                    
                    <Route path='/*' element={<AdminRoutes />} />
            }
{/*             <Route path='/auth/*' element={<AuthRoutes />} />
 */}
            <Route path='/admin/*' element={<AdminRoutes />} />
            <Route path='/employee/*' element={<EmployeeRoutes />} /> 

            <Route path='/*' element={<Navigate to='/auth' />} />

            
        </Routes>
    );
}

