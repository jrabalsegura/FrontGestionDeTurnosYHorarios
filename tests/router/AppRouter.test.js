import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { useAuthStore } from '../../src/hooks/useAuthStore';
import { AppRouter } from '../../src/router/AppRouter';

jest.mock('../../src/hooks/useAuthStore');

jest.mock('../../src/admin/pages/HomePage', () => ({
    HomePage: () => <h1>HomePage</h1>
}))

describe('Pruebas en <AppRouter />', () => {

    const mockCheckAuthToken = jest.fn();

    beforeEach(() => jest.clearAllMocks() );

    test('debe de mostrar la pantalla de carga y llamar checkAuthToken', () => {

        useAuthStore.mockReturnValue({
            status: 'checking',
            checkAuthToken: mockCheckAuthToken
        });

        render( <MemoryRouter>
                    <AppRouter />
                </MemoryRouter> )
        expect( mockCheckAuthToken ).toHaveBeenCalled();
    });

    test('debe de mostrar el login en caso de no estar autenticado', () => {
        
        useAuthStore.mockReturnValue({
            status: 'not-authenticated',
            checkAuthToken: mockCheckAuthToken
        });

        const { container } = render(
            <MemoryRouter initialEntries={['/auth2/algo/otracosa']}>
                <AppRouter />
            </MemoryRouter>
        );

        expect( screen.getByText('Accede a tu Ã¡rea privada') ).toBeTruthy();
        expect( container ).toMatchSnapshot(); 
    });

    test('debe de mostrar la homepage si estamos autenticados', () => {
        
        useAuthStore.mockReturnValue({
            status: 'authenticated',
            user: { name: 'admin', uid: '123' },
            checkAuthToken: mockCheckAuthToken
        });

        render(
            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>
        );

        expect( screen.getByText('HomePage') ).toBeTruthy();       
    });   
});