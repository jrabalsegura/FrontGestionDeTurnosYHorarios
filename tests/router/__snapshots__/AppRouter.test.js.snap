// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AppRouter /> debe de mostrar el login en caso de no estar autenticado 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-[object Object] css-1v6quvp-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item box-shadow css-hp2eq8-MuiGrid-root"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-b56s9k-MuiTypography-root"
      >
        Sistema de Gestión de Turnos y Horarios
      </h3>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item box-shadow css-8jxhq2-MuiGrid-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
      >
        Accede a tu área privada
      </h5>
      <form>
        <div
          class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1hzi63y-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                Username
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-953pxc-MuiInputBase-root-MuiInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  id=":r0:"
                  name="username"
                  placeholder="username"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              style="height: 24px;"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-2l8bkx-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r1:"
                id=":r1:-label"
              >
                Contraseña
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-953pxc-MuiInputBase-root-MuiInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  id=":r1:"
                  name="password"
                  placeholder="password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              style="height: 24px;"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1mhcw6g-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  Login
                </p>
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
